util/cluster.go

```diff
@@ -0,0 +1,205 @@
+)
+
+// GetClusterID gets the cluster ID for a provided name/ID
+// by checking in both the user cache and on the API
+func GetClusterID(clusterNameOrID string, clientWrapper *client.Wrapper) (string, error) {
+	isID := IsInClusterCache(clusterNameOrID)
+	if isID {
+		return clusterNameOrID, nil
+	}
+
+	auxParams := clientWrapper.DefaultAuxiliaryParams()
+	auxParams.ActivityName = listClustersActivityName
+
+	response, err := clientWrapper.GetClusters(auxParams)
+	if err != nil {
+		switch {
+		case clienterror.IsUnauthorizedError(err):
+			return "", microerror.Mask(errors.NotAuthorizedError)
+
+		case clienterror.IsAccessForbiddenError(err):
+			return "", microerror.Mask(errors.AccessForbiddenError)
+
+		default:
+			return "", microerror.Mask(err)
+		}
+	}
+
+	var (
+		clusterIDs    []string
+		allClusterIDs []string = make([]string, 0, len(response.Payload))
+	)
+	for _, cluster := range response.Payload {
+		allClusterIDs = append(allClusterIDs, cluster.ID)
+		if matchesValidation(clusterNameOrID, cluster) {
+			clusterIDs = append(clusterIDs, cluster.ID)
+		}
+	}
+
+	if allClusterIDs != nil {
+		CacheClusterIDs(allClusterIDs...)
+	}
+
+	switch {
+	case clusterIDs == nil:
+		return "", microerror.Mask(errors.ClusterNotFoundError)
+
+	case len(clusterIDs) > 1:
+		confirmed, id := handleNameCollision(clusterNameOrID, response.Payload)
+		if !confirmed {
+			return "", nil
+		}
+
+		return id, nil
+
+	default:
+		return clusterIDs[0], nil
+	}
+}
+
+func matchesValidation(nameOrID string, cluster *models.V4ClusterListItem) bool {
+	return cluster.DeleteDate == nil && (cluster.ID == nameOrID || cluster.Name == nameOrID)
+}
+
+func handleNameCollision(nameOrID string, clusters []*models.V4ClusterListItem) (bool, string) {
+	var (
+		clusterIDs []string
+
+		table = []string{color.CyanString("ID | ORGANIZATION | NAME")}
+	)
+
+	for _, cluster := range clusters {
+		if matchesValidation(nameOrID, cluster) {
+			clusterIDs = append(clusterIDs, cluster.ID)
+			table = append(table, fmt.Sprintf("%5s | %5s | %5s\n", cluster.ID, cluster.Owner, cluster.Name))
+		}
+	}
+
+	printNameCollisionTable(table)
+
+	confirmed, id := confirm.AskStrictOneOf(
+		fmt.Sprintf(
+			"Found more than one cluster called '%s', please type the ID of the cluster that you would like to delete",
+			nameOrID,
+		),
+		clusterIDs,
+	)
+	if !confirmed {
+		return false, ""
+	}
+	return true, id
+}
+
+func printNameCollisionTable(table []string) {
+	// Print output
+	fmt.Println("Multiple clusters found")
+	fmt.Printf("\n")
+	fmt.Println(columnize.SimpleFormat(table))
+	fmt.Printf("\n")
+}
+
+// CacheClusterIDs adds cluster IDs to a persistent cache,
+// which can be used for decreasing timeout in getting
+// cluster IDs, for commands that take both cluster names and IDs
+func CacheClusterIDs(c ...string) {
+	fs := config.FileSystem
+
+	existingC := make(chan []string)
+	go func() {
+		e, _ := readClusterCache(fs)
+		existingC <- e
+	}()
+	existing := <-existingC
```

Comment:
Why use a channel to read a file's content?

Reply:
And why read the existing cache content at all? When you cache the last cached items, why do you care what was in the cache before? This will grow the list forever, no?

Reply:
My initial thinking was that we might call this method in different places, not just when fetching all clusters. But now that you pointed it out, it probably won't get used in other places, so it doesn't really make sense to append to the existing cache (which in the first case, it should've had a limit)

Reply:
And about the channels, I wanted to somehow make cache writing non-blocking to the actual commands that use it, but I think I'm missing something in the actual implementation, because it doesn't to work like that